src/sqlfluff/utils/reflow/elements.py

```diff
@@ -0,0 +1,388 @@
+"""Dataclasses for reflow work."""
+
+
+from itertools import chain
+import logging
+from dataclasses import dataclass
+from typing import Dict, List, Optional, Sequence, Set, Tuple, cast
+
+from sqlfluff.core.parser import BaseSegment, RawSegment
+from sqlfluff.core.parser.segments.raw import WhitespaceSegment
+from sqlfluff.core.rules.base import LintFix
+
+from sqlfluff.utils.reflow.config import ReflowConfig
+from sqlfluff.utils.reflow.depthmap import DepthInfo
+
+# We're in the utils module, but users will expect reflow
+# logs to appear in the context of rules. Hence it's a subset
+# of the rules logger.
+reflow_logger = logging.getLogger("sqlfluff.rules.reflow")
+
+
+@dataclass(frozen=True)
+class ReflowElement:
+    """Base reflow element class."""
+
+    segments: Sequence[RawSegment]
+
+    @staticmethod
+    def _class_types(segments: Sequence[RawSegment]) -> Set[str]:
+        return set(chain.from_iterable(seg.class_types for seg in segments))
+
+    @property
+    def class_types(self):
+        """The set of contained class types.
+
+        Parallel to BaseSegment.class_types
+        """
+        return self._class_types(self.segments)
+
+
+@dataclass(frozen=True)
+class ReflowBlock(ReflowElement):
+    """Class for keeping track of elements to reflow.
+
+    It holds segments to reflow and also exposes configuration
+    around how they are expected to reflow around others.
+
+    The attributes exposed are designed to be "post configuration"
+    i.e. they should reflect configuration appropriately.
+
+    NOTE: These are the smallest unit of "work" within
+    the reflow methods, and may contain meta segments.
+    """
+
+    # Options for spacing rules are:
+    # - single:         the default (one single space)
+    # - close:          no whitespace
+    # - close_inline:   no whitespace and no newlines
+    # - any:            don't enforce any spacing rules
```

Comment:
This prevents reflow from being deterministic, right? This seems like a case where designing reflow for the existing rules could hold it back.

Reply:
Not necessarily but perhaps. This certainly would _allow_ multiple situations in a rule context, but I'm thinking that when we get to full reflow we might do a bit of a reset, and at that point would could take a stricter line with comments. I.e. `any` would accept multiple options but it also implies that treating it as a `single` is acceptable.

I feel like there's a more elegant solution for comments that I can't see yet. This solution feels ok for now but maybe needs a rethink further down the line. Any better ideas?

Reply:
Makes sense. I think as long as the non-any options are deterministic then it’s mission accomplished.